server/src/Irony.cpp

```diff
@@ -646,6 +639,71 @@ void Irony::getCompileOptions(const std::string &buildDir,
   std::cout << "))\n";
 
   clang_CompileCommands_dispose(compileCommands);
-  clang_CompilationDatabase_dispose(db);
 #endif
 }
+
+Irony::CDBCache::CDBCache()
+  : loaded_(false), mtime_(0) {
+}
+
+Irony::CDBCache::~CDBCache() {
+  clear();
+}
+
+std::tuple<bool, CXCompilationDatabase> Irony::CDBCache::getCompilationDatabase(const std::string &buildDir) {
+  assert(error != nullptr);
+
+  const std::string cdbFilename = buildDir + "compile_commands.json";
+
+  int ret = -2;
+
+#ifdef _WIN32
+  struct _stat st;
+  ret = _stat(cdbFilename.c_str(), &st);
+#else
+  struct stat st;
+  ret = stat(cdbFilename.c_str(), &st);
+#endif
+
+  if (ret == 0) {
+    if (cdbFilename != filename_ || st.st_mtime != mtime_) {
+      clear();
+
+      CXCompilationDatabase_Error error;
+      db_ = clang_CompilationDatabase_fromDirectory(buildDir.c_str(), &error);
+
+      switch (error) {
+      case CXCompilationDatabase_CanNotLoadDatabase:
+        std::cout << "(error . ("
+                  << "cannot-load-database"
+                  << " \"failed to load compilation database from directory\""
+                  << " " << support::quoted(buildDir) << "))\n";
+        break;
+
+      case CXCompilationDatabase_NoError:
+        loaded_ = true;
+        filename_ = cdbFilename;
+        mtime_ = st.st_mtime;
+        break;
+      }
+    }
+  }
+  else {
+    clear();
+    std::cout << "(success . ("
```

Comment:
`:success` + `"failed ..." does not look correct.

Also, I think it's better that the cache do not return error on failure to stat, only the errors from libclang should be exposed, and failure to cache should not be an error.